### ISSUE TITLE
fastlane_core/spec/command_executor_spec.rb:12 is flaky

### ISSUE THREAD

**Issue by lacostej** [fastlane/fastlane#21792]
```
  1) FastlaneCore FastlaneCore::CommandExecutor execute executes a simple command successfully
     Failure/Error: expect(result).to eq('foo')
     
       expected: "foo"
            got: ""
     
       (compared using ==)
     # ./fastlane_core/spec/command_executor_spec.rb:12:in `block (5 levels) in <top (required)>'
     # ./.bundle/ruby/3.1.0/gems/webmock-3.18.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

  2) FastlaneCore FastlaneCore::CommandExecutor execute executes a simple command successfully
     Failure/Error: expect(result).to eq('foo')
     
       expected: "foo"
            got: ""
     
       (compared using ==)
     # ./fastlane_core/spec/command_executor_spec.rb:12:in `block (5 levels) in <top (required)>'
     # ./.bundle/ruby/3.1.0/gems/webmock-3.18.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

  3) FastlaneCore FastlaneCore::CommandExecutor execute executes a simple command successfully
     Failure/Error: expect(result).to eq('foo')
     
       expected: "foo"
            got: ""
     
       (compared using ==)
     # ./fastlane_core/spec/command_executor_spec.rb:12:in `block (5 levels) in <top (required)>'
     # ./.bundle/ruby/3.1.0/gems/webmock-3.18.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Finished in 2 minutes 53.7 seconds (files took 2.1 seconds to load)
10000 examples, 3 failures
```

### New Issue Checklist

- [ ] Updated fastlane to the latest version
- [ ] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [ ] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [ ] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre> [REPLACE THIS WITH YOUR INFORMATION] </pre>
</details>

### Environment

<details>
  <pre> [REPLACE THIS WITH YOUR INFORMATION] </pre>
</details>

**Comment by lacostej** [comment]
The following script fails on some of my environments
```ruby
require 'pty'
require 'expect'

def run_command(command)
  output = []
  PTY.spawn(command) do |command_stdout, command_stdin, pid|
    begin
      command_stdout.each do |l|
        line = l.chomp
        output << line
      end
    rescue Errno::EIO
      # This is expected on some linux systems, that indicates that the subcommand finished
      # and we kept trying to read, ignore it
    ensure
      Process.wait(pid)
    end
  end
  [$?.exitstatus, output.join("\n")]
end

def run_command2(command)
  output = []
  PTY.spawn(command) do |command_stdout, command_stdin, pid|
    output = ""
    begin
      a = command_stdout.expect(/foo.*/, 5)
      output = a[0] if a
    ensure
      command_stdout.close
      command_stdin.close
      Process.wait(pid)
    end
  end
  [$?.exitstatus, output]
end

def test_spawn(command)
  status, output = run_command(command)
  errors = []
  errors << "status was '#{status}'" unless status == 0
  errors << "output was '#{output}'" unless output == "foo"
  raise errors.join(" - ") unless errors.empty?
end

t = nil
pid = nil
if ENV['STRESS']
  t = Thread.new do |t|
    puts "Spawning stress"
    pid = spawn("stress -c 16 -t 99", pgroup: true)
    puts "Waiting #{pid}"
    Process.wait(pid)
    puts "#{pid} DONE"
  end
end

command = "echo foo"
#command = "sh -c 'echo foo'"
#command = "ruby -e \"puts 'foo'\""

if ARGV.count == 1
  command = ARGV[0]
end

puts "Will run command: '#{command}'"

errors = 0
1000.times do |i|
  begin
    test_spawn(command)
  rescue => e
    puts "ERROR #{i}: #{e}"
    errors += 1
  end
end

if t
  Process.kill(:SIGKILL, -pid)
  t.join
end

raise "Failed #{errors} times" unless errors == 0
```

You can edit the script to change
* the number of iterations
* the command to test

If you install  the `stress` program, it will help adding some stress to the system to make the issues easier to reproduce.

Here are some of my findings.
```
 STRESS=y ruby test_pty.rb "echo foo"
[...]
ERROR 39: output was ''
ERROR 57: output was ''
ERROR 181: output was ''
ERROR 279: output was ''
ERROR 303: output was ''
ERROR 307: output was ''
ERROR 376: output was ''
[...]
test_pty.rb:83:in `<main>': Failed 16 times (RuntimeError)

STRESS=y ruby test_pty.rb "echo 'foo'"
[...]
ERROR 811: can't get Master/Slave device
[...]
```

so
* `echo foo` and `echo 'foo'` behave differently
* the first one reproduces the empty output rather regularly
* the second one reproduces #8258. This seems to happen only after running the test a few times, and I have a hard time to reproduce this at fresh start. Still it's a weird issue.

**Comment by lacostej** [comment]
I've reported the issue to Ruby. https://bugs.ruby-lang.org/issues/20206. Waiting for some feedback.

**Comment by lacostej** [comment]
Related to #3821 maybe?

**Comment by lacostej** [comment]
I've found out the following

running `ruby test_pty.rb "echo 'foo'"` or `ruby test_pty.rb "stdbuf -i0 -o0 -e0 echo foo"` doesn't reproduce the failure. So this could very much be caused by buffering issues on the terminal side. I still don't yet undertand how why `echo "foo"` and `echo foo` impact the buffering. I'll update the code to take this into account. Hopefully that will reduce the flakyness.

**Comment by chedabob** [comment]
@lacostej Did you manage to get any further with this? My last PR failed because of this flakey test, but I don't know enough about the underlying cause to go through and fix it.

**Comment by lacostej** [comment]
So I've a lot of different experiments and found something that works.

PR adjusted to test it on our CI infra.

Here are my experiments:

```ruby
require 'pty'
require 'expect'

class StandardError
  def exit_status
    return -1
  end
end

class FastlanePtyError < StandardError
  attr_reader :exit_status, :process_status
  def initialize(e, exit_status, process_status)
    super(e)
    set_backtrace(e.backtrace) if e
    @exit_status = exit_status
    @process_status = process_status
  end
end

class Runner
  # CLOSING - NOPE
  def run_command1(command)
    output = []
    PTY.spawn(command) do |command_stdout, command_stdin, pid|
      begin
        command_stdout.each do |l|
          line = l.chomp
          output << line
        end
      rescue Errno::EIO
        # This is expected on some linux systems, that indicates that the subcommand finished
        # and we kept trying to read, ignore it
      ensure
        begin
          command_stdout.close unless command_stdout.closed?
        rescue Errno::EIO
          # Can also be raised on close on some platforms
        end
        command_stdin.close
        #end
        Process.wait(pid)
      end
    end
    raise "#{$?.exited?} #{$?.stopped?} #{$?.signaled?} - #{$?.stopsig} - #{$?.termsig} -" unless $?.exitstatus == 0
    [$?.exitstatus, output.join("\n")]
  rescue StandardError => e
    status = $?.exitstatus
    puts $?.exitstatus if status != 0
    puts "#{$?.exited?} #{$?.stopped?} #{$?.signaled?} - #{$?.stopsig} - #{$?.termsig} -" if status != 0
    raise FastlanePtyError.new(e, $?.exitstatus || e.exit_status, status)  
  end

  # CLOSING - NOPE
  def run_command2(command)
    output = []
    PTY.spawn(command) do |command_stdout, command_stdin, pid|
      output = ""
      begin
        a = command_stdout.expect(/foo.*/, 5)
        output = a[0] if a
      ensure
        command_stdin.close
        command_stdout.close
        Process.wait(pid)
      end
    end
    raise "#{$?.exited?} #{$?.stopped?} #{$?.signaled?} - #{$?.stopsig} - #{$?.termsig} -" unless $?.exitstatus == 0
    [$?.exitstatus, output]
  end

  # NOPE
  def run_command3(command)
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    # Disable output post-processing - might reduce buffering
    command_stdout.sync = true
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    Process.wait(pid) rescue Errno::ECHILD
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # NOPE
  def run_command4(command)
    output = ""
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      loop do
        ready = IO.select([command_stdout], nil, nil, 0.1)
        break unless ready
        
        begin
          output << command_stdout.read_nonblock(4096)
        rescue IO::WaitReadable
          next
        rescue EOFError, Errno::EIO
          break
        end
      end
    ensure
      command_stdin.close unless command_stdin.closed?
      command_stdout.close unless command_stdout.closed?
      Process.wait(pid) rescue Errno::ECHILD
    end
    
    [$?.exitstatus || 0, output.strip]
  end

  # NOPE - completely broken
  def run_command5(command)
    require 'open3'
    
    # script -q /dev/null runs command in a PTY but writes to stdout
    if RUBY_PLATFORM =~ /darwin/
      wrapper = "script -q /dev/null #{command}"
    else
      wrapper = "script -q -c #{command.shellescape} /dev/null"
    end
    
    output, status = Open3.capture2e(wrapper)
    [status.exitstatus, output.strip]
  end

  require 'io/console'

  # NOPE
  def run_command6(command)
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      # Set raw mode to minimize PTY buffering
      command_stdout.raw!
    rescue Errno::ENOTTY
      # Ignore if not supported
    end
    
    begin
      # Read character-by-character or in chunks rather than line-by-line
      buffer = ""
      loop do
        begin
          buffer << command_stdout.readpartial(4096)
        rescue EOFError, Errno::EIO
          break
        end
      end
      output = buffer.split("\n").map(&:chomp)
    ensure
      command_stdin.close unless command_stdin.closed?
      command_stdout.close unless command_stdout.closed?
      Process.wait(pid) rescue Errno::ECHILD
    end
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # NOPE
  def run_command7(command)
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    output = ""
    begin
      output = command_stdout.read
    rescue Errno::EIO
      # Try to salvage any partial read
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.chomp]
  end

  # NOPE
  def run_command8(command)
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    # Give kernel time to flush PTY buffers
    sleep(0.05)
    
    # Try one more read after EIO
    begin
      while (line = command_stdout.gets)
        output << line.chomp
      end
    rescue Errno::EIO
    end
    
    Process.wait(pid) rescue Errno::ECHILD
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # HANGS
  def run_command9(command)
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    # CRITICAL: Open a second reference to the slave device
    # This prevents the kernel from cleaning up PTY buffers when child exits
    slave_path = command_stdout.path
    slave_fd = File.open(slave_path, "r") rescue nil
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
      # Expected when child exits
    ensure
      slave_fd&.close unless slave_fd&.closed?
      command_stdin.close unless command_stdin.closed?
      command_stdout.close unless command_stdout.closed?
      Process.wait(pid) rescue Errno::ECHILD
    end
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # NOPE
  def run_command10(command)
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    # Wait for process to fully exit FIRST
    Process.wait(pid)
    status = $?
    
    # NOW read - process has exited, data should be flushed
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    
    [status.exitstatus || 0, output.join("\n")]
  end

  # WORKS - first thing that works but slows down and increase CPU load under stress mode
  def run_command11(original_command)
    # Force a tiny delay before shell exits to let PTY flush
    command = "{ #{original_command}; sleep 0.00001; }"
    
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # WORKS - same CPU load issue under stress
  def run_command12(original_command)
    # runs a second command to let PTY flush
    command = "{ #{original_command}; __e=$?; :; exit $__e; }"
    
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # NOPE
  def run_command13(command)
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    # Close our write end immediately - sends EOF to child
    command_stdin.close
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # WORKS but even more complicated
  def run_command14(original_command)
    sentinel = "__EOF_#{$$}__"
    command = "{ #{original_command}; echo #{sentinel}; }"
    
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        line = l.chomp
        break if line == sentinel
        output << line
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end  

  # WORKS
  def run_command15(original_command)
    # runs a second command to let PTY flush
    command = "( #{original_command} && : )"
    
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # WORKS as well
  def run_command16(original_command)
    # runs a second command to let PTY flush
    command = "#{original_command} && :"
    
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  # WORKS
  def run_command17(original_command)
    # runs a second command to let PTY flush
    command = "#{original_command};"
    
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end

  def run_command18(original_command)
    # runs a second command to let PTY flush
    command = "(#{original_command})"
    
    output = []
    command_stdout, command_stdin, pid = PTY.spawn(command)
    
    begin
      command_stdout.each do |l|
        output << l.chomp
      end
    rescue Errno::EIO
    end
    
    command_stdin.close unless command_stdin.closed?
    command_stdout.close unless command_stdout.closed?
    Process.wait(pid) rescue Errno::ECHILD
    
    [$?.exitstatus || 0, output.join("\n")]
  end
end


def test_spawn(command, method_name)
  r = Runner.new

  status, output = r.send(method_name, command)
  errors = []
  errors << "status was '#{status}'" unless status == 0
  errors << "output was '#{output}'" unless output == "foo"
  raise errors.join(" - ") unless errors.empty?
end

t = nil
pid = nil
if ENV['STRESS']
  t = Thread.new do |t|
    puts "Spawning stress"
    pid = spawn("stress -c 16 -t 99", pgroup: true)
    puts "Waiting #{pid}"
    Process.wait(pid)
    puts "#{pid} DONE"
  end
end

command = "echo 'foo'"
#command = "sh -c 'echo foo'"
#command = "ruby -e \"puts 'foo'\""

if ARGV.count == 1
  command = ARGV[0]
end

count = 2000
COMMAND=(ENV['COMMAND'] || '1').to_i
method_name="run_command#{COMMAND}"
puts "Will run command: '#{command}' #{count} times via #{method_name}"

errors = 0
ok = 0

count.times do |i|
  begin
    test_spawn(command, method_name)
    ok += 1
  rescue => e
    puts "ERROR #{i}: #{e}"
    errors += 1
  end
end

if t && t.status
  Process.kill(:SIGKILL, -pid)
  t.join
end

raise "Failed #{errors} times" unless errors == 0
raise "Incomplete #{count-ok} times" unless ok == count
```

**Comment by lacostej** [comment]
So after many experiments, it seems that I've found an acceptable workaround. I am adding a ";" to the command which forces the flushing. There were multiple working alternatives using the same concept. (e.g. "#{command} ; :" etc). See the above script.

In the end I've remove the force closing of the streams that you can see in the script as they didn't seem to impact the final solution.

The only drawback I can think of is that it slows down a bit the process, which is not going to affect our standard usage. E.g. running the test 10000 times goes from 20s to about 2min on my Mac. 14ms per test. 

That sounds an acceptable way to remove this really really annoying behavior.

Changing the implementation would require to adjust the test.

If you want to stress test it again use the above script or checkout 67d898c10 from the branch.
